src/Database/PostgreSQL/Simple/Cursor.hs

```diff
@@ -0,0 +1,118 @@
+    , closeCursor
+    -- * Fetching rows
+    , fetchForward
+    ) where
+
+import           Data.ByteString.Builder
+                   ( Builder, byteString, char8, intDec )
+import           Control.Applicative ((<$>))
+import           Control.Exception as E
+import           Control.Monad (unless, void)
+import           Data.ByteString (ByteString)
+import           Data.Int (Int64)
+import           Data.List (intersperse)
+import           Data.Monoid (mconcat)
+import           Database.PostgreSQL.Simple.Compat ( (<>), toByteString )
+import           Database.PostgreSQL.Simple.FromField (ResultError(..))
+import           Database.PostgreSQL.Simple.FromRow (FromRow(..))
+import           Database.PostgreSQL.Simple.Ok
+import           Database.PostgreSQL.Simple.ToField (Action(..))
+import           Database.PostgreSQL.Simple.ToRow (ToRow(..))
+import           Database.PostgreSQL.Simple.Types
+                   ( Identifier(..), Query(..), (:.)(..) )
+import           Database.PostgreSQL.Simple.Internal as Base
+import           Database.PostgreSQL.Simple.Transaction
+import           Database.PostgreSQL.Simple.TypeInfo
+import qualified Database.PostgreSQL.LibPQ as PQ
+import qualified Data.ByteString.Char8 as B
+import           Control.Monad.Trans.Reader
+import           Control.Monad.Trans.State.Strict
+
+-- | Cursor.
+data Cursor = Cursor { cursorName :: !ByteString            -- ^ Escaped cursor name.
+                     , cursorConnection :: !Connection    -- ^ Connection on which the cursor was made.
+                     }
+
+-- | Declare a cursor with the given name and query
+-- on the given connection. The cursor only remains
+-- valid as long as the current transaction exists.
+declareCursor :: Connection -> ByteString -> Query -> IO Cursor
+declareCursor conn name q = do
+  name <- toByteString <$> extract <$> escapeIdentifier conn name
+  void $ execute_ conn $ mconcat ["DECLARE ", Query name, " NO SCROLL CURSOR FOR ", q]
+  return $ Cursor name conn
+  where
+    extract = either (fmtError "declareCursor" q [EscapeIdentifier name]) byteString
+
+
+-- | Declare a cursor with a temporary name and
+--
+declareTemporaryCursor :: Connection -> Query -> IO Cursor
+declareTemporaryCursor conn q = do
+  (Query name) <- newTempName conn
+  declareCursor conn name q
+
+closeCursor :: Cursor -> IO ()
+closeCursor (Cursor name conn) =
+  (void $ execute_ conn ("CLOSE " <> Query name)) `E.catch` \ex ->
+     -- Don't throw exception if CLOSE failed because the transaction is
+     -- aborted.  Otherwise, it will throw away the original error.
+     unless (isFailedTransactionError ex) $ throwIO ex
+
+-- FIXME: store "Connection" in Cursor!
+-- FIXME: ...withParser + create "plain" one with FromRow r instead
+fetchForward :: RowParser r -> Cursor -> Int -> (a -> r -> IO a) -> a -> IO a
+fetchForward parser (Cursor name conn) chunkSize f a0 = do  -- FIXME: escape name!
+  let q = toByteString (byteString "FETCH FORWARD "
```

Comment:
There's a very minor difference here wrt. the original implementation in foldWithParserBlahblah: We construct this string every time around the loop, but seeing as we're fetching from a database (thus already incurring the cost of I/O), I don't think it's a big other. Alternatively one could specify the chunkSize up front when creating the cursor, but then one loses the possibility of using an adaptive algorithm.